pkg/kubectl/cmd/util/factory.go

```diff
@@ -145,6 +145,9 @@ type Factory struct {
 	// UpdatePodSpecForObject will call the provided function on the pod spec this object supports,
 	// return false if no pod spec is supported, or return an error.
 	UpdatePodSpecForObject func(obj runtime.Object, fn func(*api.PodSpec) error) (bool, error)
+	// UpdateSelectorForObject calls fn with that applicable selector type to update.
+	UpdateSelectorForObject func(obj runtime.Object, fn func(*map[string]string, *unversioned.LabelSelector) error) error
```

Comment:
> The problem is that the old api objects use map[string]string and the newer ones use the new struct, so I need a way of returning the supported type from the api objects.

Pretty sure that `LabelSelector.MatchLabels` is the equivalent.  You should be able to detect whether you have any `.MatchExpression`.  If you don't, then its assignable.  If you do, then you fail.  Same logic either way, its just a question of whether we keep a datatype we don't want anymore in this code, right?